docs/core/forms.rst

```diff
@@ -8,278 +8,404 @@ Forms
+will assume that the value is valid.
+
+The following example shows the implementation of a custom action
+which validates that every extracted slot is valid.
+
+.. code-block:: python
+
+    from typing import Dict, Text, List, Any
+
+    from rasa_sdk import Tracker
+    from rasa_sdk.events import EventType
+    from rasa_sdk.executor import CollectingDispatcher
+    from rasa_sdk import Action
+    from rasa_sdk.events import SlotSet
+
+
+    class ValidateSlots(Action):
+        def name(self) -> Text:
+            return "action_validate_your_form"
+
+        def run(
+            self, dispatcher: CollectingDispatcher, tracker: Tracker, domain: Dict
+        ) -> List[EventType]:
+            extracted_slots: Dict[Text, Any] = tracker.get_extracted_slots()
+
+            validation_events = []
+
+            for slot_name, slot_value in extracted_slots:
+                # Check if slot is valid.
+                if is_valid(slot_value):
+                    validation_events.append(SlotSet(slot_name, slot_value))
+                else:
+                    # Return a `SlotSet` event with value `None` to indicate that this
+                    # slot still needs to be filled.
+                    validation_events.append(SlotSet(slot_name, None))
+
+            return validation_events
+
+        def is_valid(slot_value: Any) -> bool:
+            # Implementation of the validate function.
+
+
+.. _forms-custom-slot-mappings:
+
+Custom Slot Mappings
+~~~~~~~~~~~~~~~~~~~~
+
+If none of the predefined :ref:`forms-slot-mappings` fit your use case, you can use the
+:ref:`custom action<custom-actions>` ``action_validate_{form_name}`` to write your own
+extraction code. Rasa Open Source will trigger this function when the form is run.
+
+Make sure your custom action returns ``SlotSet`` events for every extracted value.
+The following example shows the implementation of a custom slot mapping which sets
+a slot based on the length of the last user message.
+
+.. code-block:: python
+
+    from typing import Dict, Text, List
+
+    from rasa_sdk import Tracker
+    from rasa_sdk.events import EventType
+    from rasa_sdk.executor import CollectingDispatcher
+    from rasa_sdk import Action
+    from rasa_sdk.events import SlotSet
+
+
+    class ValidateSlots(Action):
+        def name(self) -> Text:
+            return "action_validate_your_form"
+
+        def run(
+            self, dispatcher: CollectingDispatcher, tracker: Tracker, domain: Dict
+        ) -> List[EventType]:
+            text_of_last_user_message = tracker.latest_message.get("text")
+
+            return [SlotSet("user_message_length", len(text_of_last_user_message))]
+
+
+Requesting Extra Slots
+~~~~~~~~~~~~~~~~~~~~~~
+
+If you have frequent changes to the required slots and don't want to retrain your
+assistant when something changes, you can also use a
+:ref:`custom action<custom-actions>` ``action_validate_{form_name}`` to define
```

Comment:
did we change it to `action_validate_`? currently in the code it is `validate_`

Reply:
fixed it 👍